### PR TITLE
[LinalgToXeGPU] Support squeezable any-D memrefs

### PR DESCRIPTION
**Pull request by dchigarev**
This PR allows any-D memrefs to be used in `linalg-to-xegpu` pass as long they can be squeezed to a 2D memref through leading dimensions. Example: 
```mlir
// allowed
!chunk_type = memref<1x1x8x16xf16, strided<[512, 128, 16, 1], offset: ?>>
linalg.mul ins(%0, %1 : !chunk_type, !chunk_type) outs(%2 : !chunk_type)

// NOT allowed
!chunk_type = memref<1x8x1x16xf16, strided<[512, 128, 16, 1], offset: ?>>
linalg.mul ins(%0, %1 : !chunk_type, !chunk_type) outs(%2 : !slm_chunk)
```

The `linalg-to-xegpu` logic considers a linalg operation that can be lowered to xegpu and checks the operand shapes. The old logic refused all `linalgOps` that have non-2D operands. The new logic additionally checks if a non-2D operand can be squeezed to a 2D and allows them if they can. It then squeezes the non-2D operand using `memref.subview` and replaces old operands with new 2D memrefs. Example:
```mlir
// input:
!chunk_type = memref<1x1x8x16xf16, strided<[512, 128, 16, 1], offset: ?>>
linalg.mul ins(%0, %1 : !chunk_type, !chunk_type) outs(%2 : !chunk_type)

// output:
!chunk_type = memref<1x1x8x16xf16, strided<[512, 128, 16, 1], offset: ?>>
!squeezed_type = memref<8x16xf16, strided<[16, 1], offset: ?>>

%0_sq = memref.subview %0[0, 0, 0, 0] [1, 1, 8, 16] [1, 1, 1, 1] : !chunk_type to !squeezed_type
%1_sq = memref.subview %1[0, 0, 0, 0] [1, 1, 8, 16] [1, 1, 1, 1] : !chunk_type to !squeezed_type
%2_sq = memref.subview %2[0, 0, 0, 0] [1, 1, 8, 16] [1, 1, 1, 1] : !chunk_type to !squeezed_type
linalg.mul ins(%0_sq, %1_sq : !squeezed_type , !squeezed_type ) outs(%2_sq : !squeezed_type )
```

### CODE REVIEW

File: lib/gc/Transforms/Utils/ValueUtils.cpp [hunk 1]
```diff
@@ -193,5 +198,126 @@ bool hasSharedMemSpace(mlir::Value memref) {
...
+  while (auto subViewOp = memref.getDefiningOp<memref::SubViewOp>()) {
+    auto currentOffsets = getAsOpFoldResult(resolvedOffsets);
+    resolvedOffsets.clear();
+
+    affine::resolveIndicesIntoOpWithOffsetsAndStrides(
+        rewriter, memref.getLoc(), subViewOp.getMixedOffsets(),
+        subViewOp.getMixedStrides(), subViewOp.getDroppedDims(), currentOffsets,
+        resolvedOffsets);
+    memref = subViewOp.getOperand(0);
+  }
```
Comment (dchigarev) [inline lib/gc/Transforms/Utils/ValueUtils.cpp:217]: the logic was taken from [`XeGPUFoldAliasOps` upstream pass](https://github.com/llvm/llvm-project/blob/2b3266c1701f315d7e89c81977800001563afacb/mlir/lib/Dialect/XeGPU/Transforms/XeGPUFoldAliasOps.cpp#L45-L56)

File: include/gc/Transforms/Utils/ValueUtils.h [hunk 2]
```diff
@@ -53,6 +53,27 @@ Value flattenMemref(PatternRewriter &rewriter, Location loc, Value srcMemref);
...
 bool hasSharedMemSpace(mlir::Value memref);
 
+// Go through all parent 'memref.subview' ops for the given `memref`
+// and return the folded offsets of all subviews and the root memref.
+std::tuple<SmallVector<Value>, Value>
+computeSubviewOffsets(PatternRewriter &rewriter, Location loc, Value memref);
+
+// Return the strides of the memref
+SmallVector<OpFoldResult> getMemrefStrides(PatternRewriter &rewriter,
+                                           Location loc, Value memref);
```
Comment (AndreyPavlenko) [inline include/gc/Transforms/Utils/ValueUtils.h:63]: Maybe pass a vector reference as the function argument to avoid copying the entire vector?

Comment (dchigarev) [reply include/gc/Transforms/Utils/ValueUtils.h:63]: I think RVO is guaranteed in our case. We return the result vector straight away from `getMixedValues(...)` or `memref::computeStridesIRBlock(...)`
```c++
SmallVector<OpFoldResult> getMemrefStrides(PatternRewriter &rewriter,
                                           Location loc, Value memref) {
  auto type = dyn_cast<MemRefType>(memref.getType());

  auto stridedLayout = dyn_cast<StridedLayoutAttr>(type.getLayout());
  if (stridedLayout) {
    auto strides = stridedLayout.getStrides();
    return getMixedValues(strides, {}, rewriter);
  }

  auto sizes = getMixedValues(type.getShape(), {}, rewriter);
  auto strides = memref::computeStridesIRBlock(loc, rewriter, sizes);
  return strides;
}
```

Comment (AndreyPavlenko) [reply include/gc/Transforms/Utils/ValueUtils.h:63]: I'm not sure it's guaranteed, but I agree, in this case it does not make sense. Probably in computeSubviewOffsets() it does.

Comment (AndreyPavlenko) [reply include/gc/Transforms/Utils/ValueUtils.h:63]: I think it's a bit different case. SmallVector allocates memory on stack, you can't just move the data pointer from one vector to another, you need to copy the internal array and all the objects it contains. Compiler could optimise by inlining the function, but it's not guaranteed.

Comment (dchigarev) [reply include/gc/Transforms/Utils/ValueUtils.h]: Fixed. Now I pass the output vector to `computeSubviewOffsets()` by reference and fill it with the result

File: lib/gc/Transforms/Utils/ValueUtils.cpp [hunk 3]
```diff
@@ -155,9 +159,10 @@ Value flattenMemref(PatternRewriter &rewriter, Location loc, Value srcMemref) {
   auto srcType = cast<MemRefType>(srcMemref.getType());
 
   assert(srcType && "Expected a memref type");
-  assert(srcType.getRank() == 2 && "Expected a 2D memref");
 
-  int64_t flatSize = srcType.getShape()[0] * srcType.getShape()[1];
+  auto shapeNd = srcType.getShape();
+  int64_t flatSize =
+      std::accumulate(shapeNd.begin(), shapeNd.end(), 1, std::multiplies<>());
```
Comment (kurapov-peter) [inline lib/gc/Transforms/Utils/ValueUtils.cpp:165]: Can scalars get here somehow?

File: lib/gc/Transforms/Utils/ValueUtils.cpp [hunk 4]
```diff
@@ -193,5 +198,127 @@ bool hasSharedMemSpace(mlir::Value memref) {
   return false;
 }
 
+void computeSubviewOffsets(PatternRewriter &rewriter, Location loc,
+                           Value memref, SmallVector<Value> &resultOffsets,
+                           Value &resultRootMemref) {
+  auto fillVal = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  auto origShape = dyn_cast<MemRefType>(memref.getType()).getShape();
```
Comment (kurapov-peter) [inline lib/gc/Transforms/Utils/ValueUtils.cpp:205]: This can fail, and other places too

File: include/gc/Transforms/Utils/ValueUtils.h [hunk 5]
```diff
@@ -53,6 +53,28 @@ Value flattenMemref(PatternRewriter &rewriter, Location loc, Value srcMemref);
...
+void computeSubviewOffsets(PatternRewriter &rewriter, Location loc,
+                           Value memref, SmallVector<Value> &resultOffsets,
+                           Value &resultRootMemref);
+
+// Return the strides of the memref
+SmallVector<OpFoldResult> getMemrefStrides(PatternRewriter &rewriter,
+                                           Location loc, Value memref);
+
+// Squeeze the leading dimensions of a given memref up to 'maxDims'.
+FailureOr<Value> squeezeMemref(PatternRewriter &rewriter, Location loc,
```
Comment (kurapov-peter) [inline include/gc/Transforms/Utils/ValueUtils.h:67]: This could use a better name

File: lib/gc/Transforms/Utils/ValueUtils.cpp [hunk 6]
```diff
@@ -193,5 +198,127 @@ bool hasSharedMemSpace(mlir::Value memref) {
...
+      return rewriter.notifyMatchFailure(
+          linalgOp, "Can't squeeze memref to the desired number of dimensions");
+    }
+
+    auto flatSubview = res.value();
+    newOperands.emplace_back(i, flatSubview);
+  }
+
+  for (auto [i, operand] : newOperands)
+    linalgOp->setOperand(i, operand);
```
Comment (kurapov-peter) [inline lib/gc/Transforms/Utils/ValueUtils.cpp:307]: Does this notify the rewriter? Is it safe, in other words?

**Review by kurapov-peter — APPROVED** [review]
Some minor comments, fine otherwise

**Review by AndreyPavlenko — APPROVED** [review]

File: lib/gc/Transforms/Utils/ValueUtils.cpp (hunk 3 above)
Comment (dchigarev) [reply lib/gc/Transforms/Utils/ValueUtils.cpp:165]: nope, it's always an array:
```c++
::llvm::ArrayRef<int64_t> MemRefType::getShape() const {
  return getImpl()->shape;
}
```

File: lib/gc/Transforms/Utils/ValueUtils.cpp (hunk 4 above)
Comment (dchigarev) [reply lib/gc/Transforms/Utils/ValueUtils.cpp]: fixed

File: lib/gc/Transforms/Utils/ValueUtils.cpp (hunk 6 above)
Comment (dchigarev) [reply lib/gc/Transforms/Utils/ValueUtils.cpp]: changed to `rewriter.modifyOpInPlace`

**Pull request merged**